### DISCUSSION TITLE
How to setup persistent cache in case of programmatic API?

### DISCUSSION THREAD

**Discussion by StreetStrider** [webpack/webpack#14621] [Q&A]
Consider we have main application (A) and utility library (U). We use programmatic tool to build,which runs webpack with config object. Config object is composed from base config out of U with some optional config from A on top of it. Generated config is fed to `webpack()`.
* What paths should be added to `cache.buildDependencies` and will they be taken into account when using config object programmatically?
* How persistent cache handles imports inside config? Is it require hook when running config or AST traversing of config?
* If U is inside `node_modules/` as a package, should we depend on it, like in default we depend on `webpack/lib`?
* Can we (as an alternative), just use `cache.version` instead? Say we set `version = hash(stringify(config))`?

**Comment by alexander-akait** [comment]
> What paths should be added to cache.buildDependencies and will they be taken into account when using config object programmatically?

Which can change your output code

> If U is inside node_modules/ as a package, should we depend on it, like in default we depend on webpack/lib?
> How persistent cache handles imports inside config? Is it require hook when running config or AST traversing of config?
> Can we (as an alternative), just use cache.version instead? Say we set version = hash(stringify(config))?

We analyze your `require`/`import`

**Reply by StreetStrider** [reply]
How it is being analyzed? How it is working in case of programmatic API call? (like `webpack(config)`).

**Reply by alexander-akait** [reply]
Using this https://github.com/guybedford/es-module-lexer for ESM, and Node.js for commonjs
